src/Raven.Client/Http/FailoverNodeSelector.cs

```diff
@@ -0,0 +1,88 @@
+using System;
+using System.Threading;
+
+namespace Raven.Client.Http
+{
+    public class FailoverNodeSelector : INodeSelector
+    {
+        protected Topology _topology;
+
+        public Topology Topology => _topology;
+
+        protected int _currentNodeIndex;
+
+        public FailoverNodeSelector(Topology topology)
+        {
+            _topology = topology;
+        }
+
+        public int GetCurrentNodeIndex()
+        {
+            return _currentNodeIndex;
+        }
+
+        public virtual void OnSucceededRequest(int nodeIndex)
+        {
+            //in failover mode nothing to do
+        }
+
+        public virtual void OnFailedRequest(int nodeIndex)
+        {
+            if (Topology.Nodes.Count == 0)
+                RequestExecutor.ThrowEmptyTopology();
+
+            var nextNodeIndex = nodeIndex < Topology.Nodes.Count - 1 ? nodeIndex + 1 : 0;
+            Interlocked.CompareExchange(ref _currentNodeIndex, nextNodeIndex, nodeIndex);
+            OnNodeSwitch(nextNodeIndex);
+        }
+
+        public bool OnUpdateTopology(Topology topology, bool forceUpdate = false)
+        {
+            if (topology == null)
+                return false;
+
+            var oldTopology = _topology;
+            do
+            {
+                if (oldTopology.Etag >= topology.Etag && forceUpdate == false)
+                    return false;
+
+                if (forceUpdate == false)
+                {
+                    Interlocked.Exchange(ref _currentNodeIndex, 0);
```

Comment:
Why are you resetting the index if this isn't forced?
Also, note that you did it before you changed the topology. Existing clients may get this part way.